File: python/tests/test_structural_ops.py
import pytest
import numpy as np
from vectoria import Graph, Runtime, DType

def test_transpose_reshape():
    g = Graph()
    x_shape = [2, 3]
    perm = [1, 0]
    
    in_node = g.add_input("Input", x_shape, DType.FLOAT32)
    t_node = g.add_transpose(in_node, perm)
    r_node = g.add_reshape(t_node, [6])
    
    g.set_output(r_node)
    g.compile()
    
    rt = Runtime()
    rt.load_graph(g)
    
    x_val = np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float32)
    rt.set_input(in_node.id, x_val.flatten().tolist())
    rt.execute()
    
    out = rt.get_output(r_node.id, 6)
    out_np = np.array(out)
    
    expected = x_val.transpose(perm).reshape([6])
    
    np.testing.assert_allclose(out_np, expected, atol=1e-5)

File: core/tests/test_structural_ops.cpp
#include "vectoria/ir.hpp"
#include "vectoria/engine.hpp"
#include "vectoria/c_api.h" // For C API helpers if needed, but direct C++ API is cleaner
#include "vectoria/graph/transpose.hpp"
#include "vectoria/graph/reshape.hpp"
#include <iostream>
#include <vector>
#include <cmath>
#include <cassert>
#include <cstring>

using namespace vectoria;

void test_transpose_reshape() {
    std::cout << "Testing Structural Ops..." << std::endl;
    ir::Graph g;
    
    // Input [2, 3]
    ir::InputNode in_node;
    in_node.name = "Input";
    in_node.shape.dims = {2, 3};
    in_node.dtype = ir::DataType::Float32;
    size_t in_id = g.nodes.size();
    g.nodes.push_back({ {in_id}, in_node });
    
    // 1. Transpose [2, 3] -> [3, 2] (perm 1, 0)
    int t_id = graph::add_transpose(g, static_cast<int>(in_id), {1, 0});
    
    // 2. Reshape [3, 2] -> [6]
    int r_id = graph::add_reshape(g, t_id, {6});
    
    // 3. Reshape back to [2, 3] -> (Wait, reshape just reinterprets linear memory)
    // Transpose output is [3, 2].
    // Input:
    // [1, 2, 3]
    // [4, 5, 6]
    //
    // Transpose(1,0):
    // [1, 4]
    // [2, 5]
    // [3, 6]
    //
    // Reshape([6]):
    // [1, 4, 2, 5, 3, 6]
    
    g.outputs.push_back({static_cast<size_t>(r_id)});
    
    Engine e(g);
    e.compile();
    
    float* in_buf = static_cast<float*>(e.get_buffer(in_id));
    for(int i=0; i<6; ++i) in_buf[i] = static_cast<float>(i+1);
    
    e.execute();
    
    float* out_buf = static_cast<float*>(e.get_buffer(static_cast<size_t>(r_id)));
    
    std::vector<float> expected = {1, 4, 2, 5, 3, 6};
    
    for(int i=0; i<6; ++i) {
        if (std::abs(out_buf[i] - expected[i]) > 1e-5f) {
            std::cerr << "Structural Ops failed at index " << i << ": " << out_buf[i] << " expected " << expected[i] << std::endl;
            exit(1);
        }
    }
    
    std::cout << "Structural Ops Passed." << std::endl;
}

int main() {
    test_transpose_reshape();
    return 0;
}
